examples/step-55/step-55.cc

```diff
@@ -0,0 +1,1181 @@
+        mg_matrices.resize(0, n_levels-1);
+        mg_matrices.clear ();
+        mg_sparsity_patterns.resize(0, n_levels-1);
+
+        for (unsigned int level=0; level<n_levels; ++level)
+          {
+            DynamicSparsityPattern csp (velocity_dof_handler.n_dofs(level),
+                                        velocity_dof_handler.n_dofs(level));
+            MGTools::make_sparsity_pattern(velocity_dof_handler, csp, level);
+            mg_sparsity_patterns[level].copy_from (csp);
+
+            mg_matrices[level].reinit(mg_sparsity_patterns[level]);
+            mg_interface_matrices[level].reinit(mg_sparsity_patterns[level]);
+          }
+      }
+
+    {
+      constraints.clear ();
+      // The following makes use of a component mask for interpolation of the boundary values which
+      // is further explained in the vector valued dealii step-20 tutorial
+      DoFTools::make_hanging_node_constraints (dof_handler, constraints);
+      VectorTools::interpolate_boundary_values (dof_handler,
+                                                0,
+                                                Solution<dim>(),
+                                                constraints,
+                                                fe.component_mask(velocities));
+    }
+
+    std::vector<types::global_dof_index> dofs_per_block (2);
+    DoFTools::count_dofs_per_block (dof_handler, dofs_per_block, block_component);
+    const unsigned int n_u = dofs_per_block[0],
+                       n_p = dofs_per_block[1];
+
+    if (solver_type == SolverType::UMFPACK)
+      {
+        TimerOutput::Scope umfpack_specific(computing_timer, "(UMFPACK specific)");
+        constraints.add_line(n_u);
+      }
+    constraints.close ();
+
+    std::cout << "\tNumber of active cells: "
+              << triangulation.n_active_cells()
+              << std::endl
+              << "\tNumber of degrees of freedom: "
+              << dof_handler.n_dofs()
+              << " (" << n_u << '+' << n_p << ')'
+              << std::endl;
+
+    {
+      BlockDynamicSparsityPattern csp (2,2);
+
+      csp.block(0,0).reinit (n_u, n_u);
+      csp.block(1,0).reinit (n_p, n_u);
+      csp.block(0,1).reinit (n_u, n_p);
+      csp.block(1,1).reinit (n_p, n_p);
+
+      csp.collect_sizes();
+
+      DoFTools::make_sparsity_pattern (dof_handler, csp, constraints, false);
+      sparsity_pattern.copy_from (csp);
+
+    }
+    system_matrix.reinit (sparsity_pattern);
+
+    solution.reinit (2);
+    solution.block(0).reinit (n_u);
+    solution.block(1).reinit (n_p);
+    solution.collect_sizes ();
+
+    system_rhs.reinit (2);
+    system_rhs.block(0).reinit (n_u);
+    system_rhs.block(1).reinit (n_p);
+    system_rhs.collect_sizes ();
+  }
+
+
+// @sect4{StokesProblem::assemble_system}
+
+// In this function, the system matrix is assembled the same regardless of using ILU and GMG
+  template <int dim>
+  void StokesProblem<dim>::assemble_system ()
```

Comment:
This function looks like it's mostly copied from step-22. It would be useful if the documentation could specifically address what's different from step-22. (For example, what does `mass_factor` do?)